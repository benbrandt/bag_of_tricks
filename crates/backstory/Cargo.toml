[package]
name = "backstory"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
