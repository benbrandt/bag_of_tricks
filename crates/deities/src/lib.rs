#![deny(clippy::all)]
#![warn(clippy::pedantic)]

mod dragon;
mod dragonlance;
mod dwarven;
mod eberron;
mod elven;
mod forgotten_realms;
mod giant;
mod gnomish;
mod goblinoid;
mod greyhawk;
mod halfling;
mod historical;
mod kobold;
mod lizardfolk;
mod orc;

use std::fmt;

use alignment::{Alignment, Attitude, Morality};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::Display;

use self::{
    dragon::Dragon,
    dragonlance::Dragonlance,
    dwarven::{Duergar, Dwarven},
    eberron::Eberron,
    elven::{Drow, Elven},
    forgotten_realms::ForgottenRealms,
    giant::Giant,
    gnomish::Gnomish,
    goblinoid::{Bugbear, Goblin},
    greyhawk::Greyhawk,
    halfling::Halfling,
    historical::{Celtic, Egyptian, Greek, Norse},
    kobold::Kobold,
    lizardfolk::Lizardfolk,
    orc::Orc,
};

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Domain {
    Arcana,
    Death,
    Forge,
    Grave,
    Knowledge,
    Life,
    Light,
    Nature,
    Tempest,
    Trickery,
    War,
}

#[derive(Clone, Deserialize, Serialize)]
pub struct Deity<'a> {
    pub name: &'a str,
    pub titles: Vec<&'a str>,
    pub alignment: Alignment,
    pub domains: Vec<Domain>,
    pub symbols: Vec<&'a str>,
}

impl Deity<'_> {
    fn weight(&self, attitude_influences: &[Attitude], morality_influences: &[Morality]) -> f64 {
        self.alignment
            .weight(attitude_influences, morality_influences)
    }
}

impl<'a> fmt::Display for Deity<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "CHOSEN DEITY: {}", self.name)?;
        writeln!(f, "Titles: {}", self.titles.join("; "))?;
        writeln!(f, "Alignment: {}", self.alignment)?;
        writeln!(
            f,
            "Domains: {}",
            self.domains
                .iter()
                .map(|d| format!("{}", d))
                .collect::<Vec<_>>()
                .join(", ")
        )?;
        writeln!(f, "Symbols: {}", self.symbols.join(", "))
    }
}

pub(crate) trait Deities<'a> {
    fn deities() -> Vec<Deity<'a>>;
}

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Pantheon {
    Bugbear,
    Celtic,
    Dragon,
    Dragonlance,
    Drow,
    Duergar,
    Dwarven,
    Eberron,
    Egyptian,
    Elven,
    #[strum(serialize = "Forgotten Realms")]
    ForgottenRealms,
    Giant,
    Gnomish,
    Goblin,
    Greek,
    Greyhawk,
    Halfling,
    Kobold,
    Lizardfolk,
    Norse,
    Orc,
    None,
}

impl Pantheon {
    fn all_deities<'a>(self) -> Vec<Deity<'a>> {
        match self {
            Self::Bugbear => Bugbear::deities(),
            Self::Celtic => Celtic::deities(),
            Self::Dragon => Dragon::deities(),
            Self::Dragonlance => Dragonlance::deities(),
            Self::Drow => Drow::deities(),
            Self::Dwarven => Dwarven::deities(),
            Self::Duergar => Duergar::deities(),
            Self::Eberron => Eberron::deities(),
            Self::Egyptian => Egyptian::deities(),
            Self::Elven => Elven::deities(),
            Self::ForgottenRealms => ForgottenRealms::deities(),
            Self::Giant => Giant::deities(),
            Self::Gnomish => Gnomish::deities(),
            Self::Goblin => Goblin::deities(),
            Self::Greek => Greek::deities(),
            Self::Greyhawk => Greyhawk::deities(),
            Self::Halfling => Halfling::deities(),
            Self::Kobold => Kobold::deities(),
            Self::Lizardfolk => Lizardfolk::deities(),
            Self::Norse => Norse::deities(),
            Self::Orc => Orc::deities(),
            Self::None => vec![],
        }
    }

    #[must_use]
    pub fn deities<'a>(self, domain: Option<Domain>) -> Vec<Deity<'a>> {
        domain.map_or_else(
            || self.all_deities(),
            |d| {
                self.all_deities()
                    .into_iter()
                    .filter(|deity| deity.domains.contains(&d))
                    .collect::<Vec<_>>()
            },
        )
    }

    /// # Panics
    ///
    /// Will panic if no Pantheons available
    pub fn choose(
        rng: &mut impl Rng,
        addl_pantheons: Vec<Self>,
        domain: Option<Domain>,
        required: bool,
    ) -> Self {
        let mut options = vec![
            Self::ForgottenRealms,
            Self::Celtic,
            Self::Dragonlance,
            Self::Eberron,
            Self::Egyptian,
            Self::Greek,
            Self::Greyhawk,
            Self::Norse,
        ];
        options.extend(addl_pantheons);
        if !required {
            options.push(Self::None);
        }
        options.dedup();
        *domain
            .map_or(options.clone(), |_| {
                options
                    .into_iter()
                    .filter(|p| p == &Self::None || !p.deities(domain).is_empty())
                    .collect::<Vec<_>>()
            })
            .choose_weighted(rng, |p| p.weight())
            .unwrap()
    }

    pub fn choose_deity<'a>(
        self,
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
        domain: Option<Domain>,
    ) -> Option<Deity<'a>> {
        self.deities(domain)
            .choose_weighted(rng, |d| d.weight(attitude_influences, morality_influences))
            .ok()
            .cloned()
    }

    #[must_use]
    pub fn weight(self) -> f64 {
        match self {
            Self::Bugbear
            | Self::Dragon
            | Self::Drow
            | Self::Duergar
            | Self::Dwarven
            | Self::Elven
            | Self::ForgottenRealms
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Kobold
            | Self::Lizardfolk
            | Self::Orc
            | Self::None => 100.0,
            // Require some sort of plane shift for your character to end up in the forgotten realms
            Self::Celtic
            | Self::Dragonlance
            | Self::Eberron
            | Self::Egyptian
            | Self::Greek
            | Self::Greyhawk
            | Self::Norse => 1.0,
        }
    }
}

pub trait Pantheons {
    fn addl_pantheons(&self) -> Vec<Pantheon> {
        vec![]
    }

    fn deity_required(&self) -> bool {
        false
    }
}
