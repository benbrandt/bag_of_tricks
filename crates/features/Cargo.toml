[package]
name = "features"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
citation = { path = "../citation" }
serde = { version = "1", features = ["derive"] }

[dev-dependencies]
insta = "1"
serde_yaml = "0.8"
