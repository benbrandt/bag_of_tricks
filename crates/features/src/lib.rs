use std::fmt;

use citation::Citation;
use serde::{Deserialize, Serialize};

/// A feature or trait a character has.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Feature {
    /// Name of the feature or trait.
    pub title: &'static str,
    /// Citation for where more information about this feature is available.
    pub citation: Citation,
}

impl fmt::Display for Feature {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({})", self.title, self.citation)
    }
}

/// Trait for objects that provide features to a character.
pub trait Features {
    // Return a list of features this thing provides
    fn features(&self) -> Vec<Feature> {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use citation::Book;

    #[test]
    fn test_display() {
        let feature = Feature {
            title: "Title",
            citation: Citation(Book::Phb, 1),
        };
        insta::assert_snapshot!(format!("{}", feature));
    }
}
