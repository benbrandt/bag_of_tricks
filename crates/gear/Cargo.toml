[package]
name = "gear"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
dice_roller = { path = "../dice_roller" }
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
