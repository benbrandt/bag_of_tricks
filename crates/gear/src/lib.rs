pub mod adventuring_gear;
pub mod armor;
pub mod currency;
pub mod tools;
pub mod vehicles;
pub mod weapons;
