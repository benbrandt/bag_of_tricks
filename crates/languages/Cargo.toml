[package]
name = "languages"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
deities = { path = "../deities" }
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
