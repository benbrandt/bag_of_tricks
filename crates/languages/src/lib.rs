use deities::{Pantheon, PantheonWeight, Pantheons};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum LanguageType {
    Exotic,
    Standard,
}

impl LanguageType {
    fn weight(self) -> f64 {
        match self {
            Self::Exotic => 1.0,
            Self::Standard => 2.0,
        }
    }
}

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gith,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub fn gen(
        rng: &mut impl Rng,
        current_languages: &[Language],
        (amount, language_type): (usize, Option<LanguageType>),
    ) -> Vec<Self> {
        language_type
            .map_or_else(
                || Self::iter().collect::<Vec<Self>>(),
                |lt| {
                    let filtered = Self::iter()
                        .filter(|l| l.language_type() == lt)
                        .collect::<Vec<Self>>();
                    if filtered.is_empty() {
                        Self::iter().collect::<Vec<Self>>()
                    } else {
                        filtered
                    }
                },
            )
            .iter()
            .filter(|l| !current_languages.contains(l))
            .copied()
            .collect::<Vec<_>>()
            .choose_multiple_weighted(rng, amount, |l| l.language_type().weight())
            .unwrap()
            .copied()
            .collect()
    }

    fn language_type(self) -> LanguageType {
        match self {
            Self::Common
            | Self::Dwarvish
            | Self::Elvish
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Orc => LanguageType::Standard,
            Self::Abyssal
            | Self::Celestial
            | Self::DeepSpeech
            | Self::Draconic
            | Self::Gith
            | Self::Infernal
            | Self::Primordial
            | Self::Sylvan
            | Self::Undercommon => LanguageType::Exotic,
        }
    }
}

impl Pantheons for Language {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        (match self {
            Language::Abyssal
            | Language::Celestial
            | Language::DeepSpeech
            | Language::Gith
            | Language::Infernal
            | Language::Primordial
            | Language::Sylvan => vec![],
            Language::Common => vec![Pantheon::ForgottenRealms],
            Language::Draconic => vec![Pantheon::Dragon, Pantheon::Kobold, Pantheon::Lizardfolk],
            Language::Dwarvish => vec![Pantheon::Dwarven],
            Language::Elvish => vec![Pantheon::Elven],
            Language::Giant => vec![Pantheon::Giant],
            Language::Gnomish => vec![Pantheon::Gnomish],
            Language::Goblin => vec![Pantheon::Bugbear, Pantheon::Goblin],
            Language::Halfling => vec![Pantheon::Halfling],
            Language::Orc => vec![Pantheon::Orc],
            Language::Undercommon => vec![Pantheon::Drow, Pantheon::Duergar],
        })
        .into_iter()
        .map(|p| (p, PantheonWeight::Possible))
        .collect::<Vec<_>>()
    }
}

/// Trait for describing what languages are provided by a given object.
pub trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (0, None)
    }
}
