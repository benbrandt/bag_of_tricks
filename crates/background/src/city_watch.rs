use std::fmt;

use backstory::Backstory;
use citation::{Book, Citation, CitationList, Citations};
use deities::Pantheons;
use features::{Feature, Features};
use gear::{
    adventuring_gear::{Gear, OtherGear},
    currency::Coin,
};
use languages::{LanguageType, Languages};
use personality::{Influence, PersonalityOptions};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use stats::{
    ability::{AbilityScores, Skill},
    equipment::{Equipment, StartingEquipment},
    proficiencies::{Proficiencies, Proficiency},
};
use strum::{Display, EnumIter, IntoEnumIterator};

use super::{
    max_skill_weight,
    soldier::{BONDS, FLAWS, IDEALS, TRAITS},
    Background,
};

#[derive(Copy, Clone, Deserialize, Display, EnumIter, Serialize)]
enum Variant {
    #[strum(serialize = "City Watch")]
    CityWatch,
    Investigator,
}

impl Default for Variant {
    fn default() -> Self {
        Self::CityWatch
    }
}

impl Variant {
    fn gen(
        rng: &mut impl Rng,
        ability_scores: &AbilityScores,
        proficiencies: &[Proficiency],
        proficiency_bonus: i16,
    ) -> Self {
        *Self::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |v| {
                v.weight(ability_scores, proficiencies, proficiency_bonus)
            })
            .unwrap()
    }

    fn skills(self) -> Vec<Skill> {
        match self {
            Variant::CityWatch => vec![Skill::Athletics, Skill::Insight],
            Variant::Investigator => vec![Skill::Insight, Skill::Investigation],
        }
    }

    fn weight(
        self,
        ability_scores: &AbilityScores,
        proficiencies: &[Proficiency],
        proficiency_bonus: i16,
    ) -> f64 {
        max_skill_weight(
            &self.skills(),
            ability_scores,
            proficiencies,
            proficiency_bonus,
        )
    }
}

#[derive(Default, Deserialize, Serialize)]
pub struct CityWatch {
    variant: Variant,
}

impl Background for CityWatch {
    fn gen(
        rng: &mut impl Rng,
        ability_scores: &AbilityScores,
        proficiencies: &[Proficiency],
        proficiency_bonus: i16,
    ) -> Self {
        Self {
            variant: Variant::gen(rng, ability_scores, proficiencies, proficiency_bonus),
        }
    }

    fn skills() -> Vec<Skill> {
        vec![Skill::Athletics, Skill::Insight, Skill::Investigation]
    }

    fn weight(
        ability_scores: &AbilityScores,
        proficiencies: &[Proficiency],
        proficiency_bonus: i16,
    ) -> f64 {
        Variant::iter()
            .map(|v| v.weight(ability_scores, proficiencies, proficiency_bonus))
            .max_by(|a, b| a.partial_cmp(b).unwrap())
            .unwrap_or(0.0)
    }
}

impl Backstory for CityWatch {}

impl Citations for CityWatch {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Scag, 145)])
    }
}

impl Features for CityWatch {
    fn features(&self) -> Vec<Feature> {
        // Your experience in enforcing the law, and dealing with lawbreakers, gives you a feel for local laws and criminals. You can easily find the local outpost of the watch or a similar organization, and just as easily pick out the dens of criminal activity in a community, although you're more likely to be welcome in the former locations rather than the latter.
        vec![Feature {
            title: "Watcher's Eye",
            citation: Citation(Book::Scag, 145),
        }]
    }
}

impl Languages for CityWatch {
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (2, None)
    }
}

impl Pantheons for CityWatch {}

impl PersonalityOptions for CityWatch {
    fn bonds(&self) -> Vec<String> {
        BONDS.iter().map(|&s| s.to_string()).collect()
    }

    fn flaws(&self) -> Vec<String> {
        FLAWS.iter().map(|&s| s.to_string()).collect()
    }

    fn ideals(&self) -> Vec<(String, Influence)> {
        IDEALS.iter().map(|&(s, i)| (s.to_string(), i)).collect()
    }

    fn traits(&self) -> Vec<String> {
        TRAITS.iter().map(|&s| s.to_string()).collect()
    }
}

impl Proficiencies for CityWatch {
    fn proficiencies(&self) -> Vec<Proficiency> {
        self.variant
            .skills()
            .iter()
            .map(|&s| Proficiency::Skill(s))
            .collect()
    }
}

impl StartingEquipment for CityWatch {
    fn coins(&self) -> (Coin, u8) {
        (Coin::Gold, 10)
    }

    fn equipment(&self) -> Vec<Equipment> {
        vec![
            Equipment::Other(
                "A uniform in the style of your unit and indicative of your rank".into(),
            ),
            Equipment::Other("a horn with which to summon help".into()),
            Equipment::Gear(Gear::Other(OtherGear::Manacles)),
            Equipment::Gear(Gear::Other(OtherGear::Pouch)),
        ]
    }
}

impl fmt::Display for CityWatch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.variant)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_display_snapshot!(background);
    }

    #[test]
    fn test_bonds() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.bonds());
    }

    #[test]
    fn test_flaws() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.flaws());
    }

    #[test]
    fn test_ideals() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.ideals());
    }

    #[test]
    fn test_traits() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = CityWatch::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.traits());
    }

    #[test]
    fn test_skills() {
        insta::assert_yaml_snapshot!(CityWatch::skills());
    }

    #[test]
    fn test_weight() {
        insta::assert_yaml_snapshot!(CityWatch::weight(&AbilityScores::default(), &[], 2));
    }

    #[test]
    fn test_snapshot_citations() {
        let background = CityWatch {
            variant: Variant::CityWatch,
        };
        insta::assert_yaml_snapshot!(background.citations());
    }

    #[test]
    fn test_snapshot_features() {
        let background = CityWatch {
            variant: Variant::CityWatch,
        };
        insta::assert_yaml_snapshot!(background.features());
    }

    #[test]
    fn test_snapshot_languages() {
        let background = CityWatch {
            variant: Variant::CityWatch,
        };
        insta::assert_yaml_snapshot!(background.addl_languages());
    }

    #[test]
    fn test_snapshot_proficiences() {
        insta::assert_yaml_snapshot!(Variant::iter()
            .map(|variant| CityWatch { variant }.proficiencies())
            .collect::<Vec<Vec<Proficiency>>>());
    }

    #[test]
    fn test_snapshot_coins() {
        let background = CityWatch {
            variant: Variant::CityWatch,
        };
        insta::assert_yaml_snapshot!(background.coins());
    }

    #[test]
    fn test_snapshot_equipment() {
        let background = CityWatch {
            variant: Variant::CityWatch,
        };
        insta::assert_yaml_snapshot!(background.equipment());
    }
}
