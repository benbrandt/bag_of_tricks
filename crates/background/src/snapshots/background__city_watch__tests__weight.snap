---
source: crates/background/src/city_watch.rs
expression: "CityWatch::weight(&AbilityScores::default(), &[], 2)"

---
1.0

